webhooks/virtualmachinesnapshot/validation/virtualmachinesnapshot_validator.go

```diff
@@ -56,22 +61,85 @@ func (v validator) For() schema.GroupVersionKind {
 	return vmopv1.GroupVersion.WithKind(reflect.TypeOf(vmopv1.VirtualMachineSnapshot{}).Name())
 }
 
+// ValidateCreate makes sure the VirtualMachineSnapshot create request is valid.
 func (v validator) ValidateCreate(ctx *pkgctx.WebhookRequestContext) admission.Response {
-	// TODO: Implement validation for create
+	vmSnapshot, err := v.vmSnapshotFromUnstructured(ctx.Obj)
+	if err != nil {
+		return webhook.Errored(http.StatusBadRequest, err)
+	}
 
 	validationErrs := make([]string, 0)
 
+	if vmSnapshot.Spec.VMRef == nil {
+		validationErrs = append(validationErrs, "spec.vmRef must not be nil")
+	}
+
+	if vmSnapshot.Spec.VMRef != nil {
+		if vmSnapshot.Spec.VMRef.Kind != virtualMachineKind {
+			validationErrs = append(validationErrs, fmt.Sprintf("spec.vmRef.kind must be %q, got %q",
+				virtualMachineKind, vmSnapshot.Spec.VMRef.Kind))
+		}
+
+		if vmSnapshot.Spec.VMRef.APIVersion == "" {
+			validationErrs = append(validationErrs, "spec.vmRef.apiVersion must be set to a non-empty value")
+		}
+
+		if vmSnapshot.Spec.VMRef.Name == "" {
+			validationErrs = append(validationErrs, "spec.vmRef.name must be set to a non-empty value")
+		} else {
+			// in addition to fetching the VM referred to in the snapshot, it also
```

Comment:
@aruneshpa Can we please have our validation webhooks just focus on internal consistency? There is no guarantee that any of this will be true when the snapshot is actually reconciled.

Reply:
True that the guarantee might not hold good when the object is reconciled.

But this protects against the case where a VM is powered off and the user is requesting for a snapshot with memory included. User gets an early warning and ensures that an unnecessary VMSnapshot resource isn't created.

_Unless_ the behavior we want is that users can create a VMSnapshot with memory=true even when VM=PoweredOff **and** if and when the VM is powered back on, the snapshot is taken.

Reply:
It really doesn't protect against anything. Status can be stale now or later. Any errors need to be reported in a condition, users need to have a done/ready/whatever condition to wait/check for completion.